lms/views/api/blackboard/sync.py

```diff
@@ -11,68 +12,69 @@ def __init__(self, request):
+        group_info = params["group_info"]
+        grading_student_id = params["data"].get("gradingStudentId")
+
+        group_set_id = self.group_set(tool_consumer_instance_guid, resource_link_id)
+        course = self.get_course(
+            tool_consumer_instance_guid, params["course"]["context_id"]
+        )
 
-        self.sync_to_h(groups)
+        groups = self.get_blackboard_groups(course, group_set_id, grading_student_id)
 
+        self.request.find_service(name="lti_h").sync(groups, group_info)
         authority = self.request.registry.settings["h_authority"]
         return [group.groupid(authority) for group in groups]
 
-    def get_blackboard_groups(self):
+    def get_blackboard_groups(self, course, group_set_id, grading_student_id=None):
         lti_user = self.request.lti_user
-        group_set_id = self.group_set()
-        course_id = self.request.json["course"]["context_id"]
-        course = self.get_course(course_id)
 
         if lti_user.is_learner:
             user = self.request.find_service(UserService).get(
-                self.request.find_service(name="application_instance").get_current(),
+                course.application_instance,
                 lti_user.user_id,
             )
 
             learner_groups = self.blackboard_api.course_groups(
-                course_id, group_set_id, current_student_own_groups_only=True
+                course.lms_id, group_set_id, current_student_own_groups_only=True
             )
             groups = self.to_groups_groupings(course, learner_groups)
             self.grouping_service.upsert_grouping_memberships(user, groups)
             return groups
 
-        if grading_student_id := self.request.json.get("gradingStudentId"):
+        if grading_student_id:
             return self.grouping_service.get_course_groupings_for_user(
                 course,
                 grading_student_id,
                 type_=Grouping.Type.BLACKBOARD_GROUP,
                 group_set_id=group_set_id,
             )
 
-        groups = self.blackboard_api.group_set_groups(course_id, group_set_id)
+        groups = self.blackboard_api.group_set_groups(course.lms_id, group_set_id)
         return self.to_groups_groupings(course, groups)
 
-    def group_set(self):
+    def group_set(self, tool_consumer_instance_guid, resource_link_id):
         return (
             self.request.find_service(name="assignment")
-            .get(
-                self.tool_consumer_instance_guid,
-                self.request.json["assignment"]["resource_link_id"],
-            )
+            .get(tool_consumer_instance_guid, resource_link_id)
             .extra["group_set_id"]
         )
 
     def to_groups_groupings(self, course, groups):
         return [
             self.grouping_service.upsert_with_parent(
-                tool_consumer_instance_guid=self.tool_consumer_instance_guid,
+                tool_consumer_instance_guid=course.application_instance.tool_consumer_instance_guid,
```

Comment:
`upsert_with_parent` might not need to take `tool_consumer_instance_guid` at all and do this instead but I reckon that would be better on a separate PR.